src/uproot/interpretation/objects.py

```diff
@@ -437,29 +437,54 @@ def final_array(
         )
         trimmed = []
         start = entry_offsets[0]
+        has_any_awkward_types = any(
+            uproot._util.from_module(x, "awkward") for x in basket_arrays.values()
+        )
         for basket_num, stop in enumerate(entry_offsets[1:]):
+            to_append = None
             if start <= entry_start and entry_stop <= stop:
                 local_start = entry_start - start
                 local_stop = entry_stop - start
-                trimmed.append(basket_arrays[basket_num][local_start:local_stop])
+                to_append = basket_arrays[basket_num][local_start:local_stop]
 
             elif start <= entry_start < stop:
                 local_start = entry_start - start
                 local_stop = stop - start
-                trimmed.append(basket_arrays[basket_num][local_start:local_stop])
+                to_append = basket_arrays[basket_num][local_start:local_stop]
 
             elif start <= entry_stop <= stop:
                 local_start = 0
                 local_stop = entry_stop - start
-                trimmed.append(basket_arrays[basket_num][local_start:local_stop])
+                to_append = basket_arrays[basket_num][local_start:local_stop]
 
             elif entry_start < stop and start <= entry_stop:
-                trimmed.append(basket_arrays[basket_num])
+                to_append = basket_arrays[basket_num]
+
+            if (
+                to_append is not None
+                and isinstance(
+                    library,
+                    (
+                        uproot.interpretation.library.Awkward,
+                        uproot.interpretation.library.Pandas,
+                    ),
+                )
+                and isinstance(to_append, numpy.ndarray)
+                and has_any_awkward_types
+            ):
+                trimmed.append(
+                    uproot.interpretation.library._object_to_awkward_array(
+                        uproot.extras.awkward(), self._form, to_append
+                    )
+                )
+            elif to_append is not None:
+                trimmed.append(to_append)
```

Comment:
In the proposed change:
  if we were to have library = ak and only NumPy types, then it will go into the if clause and turn all the NumPy types into Awkward types by relying on a form that doesn’t exist; instead of simply appending all the NumPy types into `trimmed` which will then get concatenated and transformed into Awkward through line:  https://github.com/scikit-hep/uproot5/blob/1cee9606748482c8a06828646144bd44caac4b8f/src/uproot/interpretation/objects.py#L487